### DISCUSSION TITLE
RuntimeError in Custom PyTorch Model during Backpropagation

### DISCUSSION THREAD

**Discussion by meghnaSunil** [community/community#98159] [Programming Help]
Heyy I am getting this error and need some help figuring it out

RuntimeError: one of the variables needed for gradient computation has been modified by an inplace operation

### Question:

I'm implementing a custom neural network model in PyTorch for a multi-task learning problem and encountered a runtime error during the backpropagation step. The model has a common set of layers for feature extraction and separate branches for each task. The error occurs during backpropagation:

Here's the simplified code structure:

```python
import torch
import torch.nn as nn
import torch.optim as optim

class CustomModel(nn.Module):
    def _init_(self, in_features, task1_out_features, task2_out_features):
        super(CustomModel, self)._init_()
        # Shared layers
        self.shared_layers = nn.Sequential(
            nn.Linear(in_features, 128),
            nn.ReLU(),
            nn.Linear(128, 256),
            nn.ReLU()
        )
        # Task-specific layers
        self.task1_head = nn.Linear(256, task1_out_features)
        self.task2_head = nn.Linear(256, task2_out_features)

    def forward(self, x):
        x = self.shared_layers(x)
        out1 = self.task1_head(x)
        out2 = self.task2_head(x)
        return out1, out2

# Model, loss function, optimizer initialization
in_features = 64 # Example feature size
task1_out_features = 10 # Example output size for task 1
task2_out_features = 5 # Example output size for task 2
batch_size = 32 # Example batch size

model = CustomModel(in_features, task1_out_features, task2_out_features)
criterion1 = nn.CrossEntropyLoss()
criterion2 = nn.MSELoss()
optimizer = optim.Adam(model.parameters(), lr=0.001)

# Dummy data for demonstration
input_data = torch.randn(batch_size, in_features)
target1 = torch.empty(batch_size, dtype=torch.long).random_(task1_out_features)
target2 = torch.randn(batch_size, task2_out_features)

# Forward pass
output1, output2 = model(input_data)

# Loss calculation
loss1 = criterion1(output1, target1)
loss2 = criterion2(output2, target2)
total_loss = loss1 + loss2

# Backward pass
optimizer.zero_grad()
total_loss.backward()
optimizer.step()

**Comment by jkoda-rsa** [comment]
### Answer:

The error you're encountering is often caused by modifying a tensor in-place in a way that interferes with the gradient computation. In PyTorch, certain in-place operations on tensors can overwrite values needed to compute gradients, leading to this error during the backward pass.

In your code, the likely culprit is not immediately apparent, as there are no explicit in-place operations in the forward method of your `CustomModel`. However, the issue might be in other parts of your code not shown here or in the hidden behavior of some modules or methods you've used.

Maybe this could resolve your issue:

- **Check for Hidden In-Place Operations**: Review all parts of your code to ensure there are no in-place operations (like `x += y` or `x[:] = y`) that might affect the tensors involved in your model's computations.

- **Update Batch Normalization Layers**: If you have batch normalization layers (not shown in your code snippet), ensure they are not causing the issue. Sometimes, batch normalization can lead to such errors under specific configurations.

Hope it helps!

Cheers,
Jenestin

**Comment by LuluW8071** [comment]
> RuntimeError: one of the variables needed for gradient computation has been modified by an inplace operation

Try  this snippet 
```python 
import torch
import torch.nn as nn

class CustomModel(nn.Module):
    def __init__(self, in_features, task1_out_features, task2_out_features):
        super(CustomModel, self).__init__()
        # Shared layers
        self.shared_layers = nn.Sequential(
            nn.Linear(in_features, 128),
            nn.ReLU(),
            nn.Linear(128, 256),
            nn.ReLU()
        )
        # Task-specific layers
        self.task1_head = nn.Linear(256, task1_out_features)
        self.task2_head = nn.Linear(256, task2_out_features)

    def forward(self, x):
        x = self.shared_layers(x)
        out1 = self.task1_head(x)
        out2 = self.task2_head(x)
        return out1, out2

# Model, loss function, optimizer initialization
in_features = 64 # Example feature size
task1_out_features = 10 # Example output size for task 1
task2_out_features = 5 # Example output size for task 2
batch_size = 32 # Example batch size

model = CustomModel(in_features, task1_out_features, task2_out_features)
criterion1 = nn.CrossEntropyLoss()
criterion2 = nn.MSELoss()
optimizer = torch.optim.Adam(model.parameters(), lr=0.001)

# Dummy data for demonstration
input_data = torch.randn(batch_size, in_features)
target1 = torch.empty(batch_size, dtype=torch.long).random_(task1_out_features)
target2 = torch.randn(batch_size, task2_out_features)

# Forward pass
output1, output2 = model(input_data)

# Loss calculation
loss1 = criterion1(output1, target1)
loss2 = criterion2(output2, target2)
total_loss = loss1 + loss2

# Backward pass
optimizer.zero_grad()
total_loss.backward()
optimizer.step()

```

Use double underscore before and after for the constructor `__init__()`
This resolves the calling instance of model resolving `inplace operation` error
